pkg/controller/controller_test.go

```diff
@@ -635,6 +590,321 @@ func TestRolloutController_Reconcile(t *testing.T) {
+			scheme.AddKnownTypeWithName(customResourceGVK, &dummy{})
+			restMapper := testrestmapper.TestOnlyStaticRESTMapper(scheme)
+
+			kubeClient := fake.NewSimpleClientset(testData.statefulSets...)
+
+			// Inject a hook to track all patched StatefulSets or return mocked errors.
+			patchedStsNames := addPatchStatefulsetReactor(kubeClient, testData.kubePatchErr)
+			scaleClient := createScaleClient(testData.customResourceScaleSpecReplicas, testData.customResourceScaleStatusReplicas, testData.getScaleErr)
+
+			dynamicClient, _ := createFakeDynamicClient()
+			httpClient := &fakeHttpClient{
+				responses:       testData.httpResponses,
+				defaultResponse: internalErrorResponse,
+			}
+
+			// Create the controller and start informers.
+			reg := prometheus.NewPedanticRegistry()
+			c := NewRolloutController(kubeClient, restMapper, scaleClient, dynamicClient, testNamespace, httpClient, 5*time.Second, reg, log.NewNopLogger())
+			require.NoError(t, c.Init())
+			defer c.Stop()
+
+			// Run a reconcile.
+			require.NoError(t, c.reconcile(context.Background()))
+
+			// Assert patched StatefulSets.
+			assert.Equal(t, testData.expectedPatchedSets, convertEmptyMapToNil(patchedStsNames))
+			assert.ElementsMatch(t, testData.expectedHttpRequests, httpClient.requests())
+		})
+	}
+}
+
+func createFakeDynamicClient() (*fakedynamic.FakeDynamicClient, map[string][]string) {
+	dynamicClient := &fakedynamic.FakeDynamicClient{}
+
+	patchedStatuses := map[string][]string{}
+	dynamicClient.AddReactor("patch", "*", func(action ktesting.Action) (handled bool, ret runtime.Object, err error) {
+		patchAction := action.(ktesting.PatchAction)
+		name := path.Join(patchAction.GetResource().Group, patchAction.GetResource().Version, patchAction.GetResource().Resource, patchAction.GetName(), patchAction.GetSubresource())
+		if patchedStatuses == nil {
+			patchedStatuses = map[string][]string{}
+		}
+		patchedStatuses[name] = append(patchedStatuses[name], string(action.(ktesting.PatchAction).GetPatch()))
+		return true, nil, nil
+	})
+	return dynamicClient, patchedStatuses
+}
+
+// addPatchStatefulsetReactor injects a hook to track all patched StatefulSets or return mocked errors.
+func addPatchStatefulsetReactor(fakeKubeClient *fake.Clientset, kubePatchErr error) map[string][]string {
+	patchedStsNames := map[string][]string{}
+	fakeKubeClient.PrependReactor("patch", "*", func(action ktesting.Action) (handled bool, ret runtime.Object, err error) {
+		if kubePatchErr != nil {
+			return true, nil, kubePatchErr
+		}
+
+		switch action.GetResource().Resource {
+		case "statefulsets":
+			name := action.(ktesting.PatchAction).GetName()
+			patchedStsNames[name] = append(patchedStsNames[name], string(action.(ktesting.PatchAction).GetPatch()))
+		}
+
+		return false, nil, nil
+	})
+	return patchedStsNames
+}
+
+func createScaleClient(customResourceScaleSpecReplicas int, customResourceScaleStatusReplicas int, getScaleErr error) *fakescale.FakeScaleClient {
```

Comment:
[nit] To keep it consistent with `createFakeDynamicClient()`

```suggestion
func createFakeScaleClient(customResourceScaleSpecReplicas int, customResourceScaleStatusReplicas int, getScaleErr error) *fakescale.FakeScaleClient {
```